File: runtime/cpp/titan-ast-runtime-lib/ByteBuffer.cpp
//
// Created by tian wei jun on 2022/11/27 0027.
//

#include "ByteBuffer.h"

ByteBuffer::ByteBuffer(int capacity)
    : isBigEndian(true), capacity(capacity), position(0) {
  buffer = new byte[capacity];
}

ByteBuffer::ByteBuffer(int capacity, bool isBigEndian)
    : isBigEndian(isBigEndian), capacity(capacity), position(0) {
  buffer = new byte[capacity];
}

ByteBuffer::ByteBuffer(const ByteBuffer &byteBuffer) {
  this->isBigEndian = byteBuffer.isBigEndian;
  this->capacity = byteBuffer.capacity;
  this->position = byteBuffer.position;
  this->buffer = new byte[this->capacity];
  for (int i = 0; i < this->capacity; i++) {
    this->buffer[i] = byteBuffer.buffer[i];
  }
}

ByteBuffer::ByteBuffer(ByteBuffer &&byteBuffer) {
  this->isBigEndian = byteBuffer.isBigEndian;
  this->capacity = byteBuffer.capacity;
  this->position = byteBuffer.position;
  this->buffer = byteBuffer.buffer;

  byteBuffer.buffer = nullptr;
  byteBuffer.capacity = 0;
  byteBuffer.position = 0;
}

ByteBuffer::~ByteBuffer() {
  delete[] buffer;
  buffer = nullptr;
}

void ByteBuffer::setPosition(int position) { this->position = position; }

int ByteBuffer::length() const { return position; }

void ByteBuffer::append(byte b) {
  if (position >= capacity) {
    extendBuffer();
  }
  buffer[position++] = b;
}

void ByteBuffer::extendBuffer() {
  int newCapacity = capacity + capacity;
  byte *newBuffer = new byte[newCapacity];
  for (int i = 0; i < capacity; i++) {
    newBuffer[i] = buffer[i];
  }
  delete[] buffer;
  buffer = newBuffer;
  capacity = newCapacity;
}

void ByteBuffer::clear() { position = 0; }

int ByteBuffer::getInt() const { return isBigEndian ? getIntB() : getIntL(); }

int ByteBuffer::getIntB() const {
  int base = 0;
  int value = 0;
  for (int i = position - 1; i >= 0; i--) {
    byte tmp = buffer[i];
    value = value | tmp << base;
    base += 8;
  }
  return value;
}

int ByteBuffer::getIntL() const {
  int base = 0;
  int value = 0;
  for (int i = 0; i < position; i++) {
    byte tmp = buffer[i];
    value = value | tmp << base;
    base += 8;
  }
  return value;
}

File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/ByteBuffer.java
package titan.ast.runtime;

import java.nio.charset.StandardCharsets;

/**
 * .
 *
 * @author tian wei jun
 */
public class ByteBuffer {
  boolean isBigEndian = true;
  int position = 0;
  byte[] buffer;

  public ByteBuffer(int capacity) {
    this.buffer = new byte[capacity];
  }

  void setPosition(int position) {
    this.position = position;
  }

  int length() {
    return position;
  }

  void append(int b) {
    if (position >= buffer.length) {
      extendBuffer();
    }
    buffer[position++] = (byte) (b & 0xFF);
  }

  void append(byte b) {
    if (position >= buffer.length) {
      extendBuffer();
    }
    buffer[position++] = b;
  }

  void extendBuffer() {
    int newCapacity = buffer.length + buffer.length;
    byte[] newBuffer = new byte[newCapacity];
    System.arraycopy(buffer, 0, newBuffer, 0, buffer.length);
    buffer = newBuffer;
  }

  void clear() {
    position = 0;
  }

  int getInt() {
    return isBigEndian ? getIntB() : getIntL();
  }

  int getIntB() {
    int base = 0;
    int value = 0;
    for (int i = position - 1; i >= 0; i--) {
      byte tmp = buffer[i];
      value = value | tmp << base;
      base += 8;
    }
    return value;
  }

  int getIntL() {
    int base = 0;
    int value = 0;
    for (int i = 0; i < position; i++) {
      byte tmp = buffer[i];
      value = value | tmp << base;
      base += 8;
    }
    return value;
  }

  @Override
  public String toString() {
    return new String(buffer, 0, position, StandardCharsets.ISO_8859_1);
  }
}
